executor/index_lookup_hash_join.go

```diff
@@ -482,3 +592,64 @@ func (iw *indexHashJoinInnerWorker) joinMatchedInnerRow2Chunk(ctx context.Contex
 	}
 	return true, joinResult
 }
+
+func (iw *indexHashJoinInnerWorker) collectMatchedInnerPtrs4OuterRows(ctx context.Context, innerRow chunk.Row, innerRowPtr chunk.RowPtr, task *indexHashJoinTask,
+	joinResult *indexHashJoinResult, h hash.Hash64, buf []byte) error {
+	_, matchedOuterRowIdx, err := iw.getMatchedOuterRows(innerRow, task, h, buf)
+	if err != nil {
+		return err
+	}
+	for _, outerRowIdx := range matchedOuterRowIdx {
+		task.matchedInnerRowPtrs[outerRowIdx] = append(task.matchedInnerRowPtrs[outerRowIdx], innerRowPtr)
+	}
+	return nil
+}
+
+// doJoinInOrder follows the following steps:
+// 1. collect all the matched inner row ptrs for every outer row
+// 2. do the join work
+//   2.1 collect all the matched inner rows using the collected ptrs for every outer row
+//   2.2 call tryToMatchInners for every outer row
+//   2.3 call onMissMatch when no inner rows are matched
+func (iw *indexHashJoinInnerWorker) doJoinInOrder(ctx context.Context, task *indexHashJoinTask, joinResult *indexHashJoinResult, h hash.Hash64, resultCh chan *indexHashJoinResult) (err error) {
+	for i, numChunks := 0, task.innerResult.NumChunks(); i < numChunks; i++ {
+		for j, chk := 0, task.innerResult.GetChunk(i); j < chk.NumRows(); j++ {
+			row := chk.GetRow(j)
+			ptr := chunk.RowPtr{ChkIdx: uint32(i), RowIdx: uint32(j)}
+			err = iw.collectMatchedInnerPtrs4OuterRows(ctx, row, ptr, task, joinResult, h, iw.joinKeyBuf)
+		}
+	}
+	// TODO: matchedInnerRowPtrs and matchedInnerRows can be moved to inner worker.
+	matchedInnerRows := make([]chunk.Row, len(task.matchedInnerRowPtrs))
+	var hasMatched, hasNull, ok bool
+	for outerRowIdx, innerRowPtrs := range task.matchedInnerRowPtrs {
+		matchedInnerRows, hasMatched, hasNull = matchedInnerRows[:0], false, false
+		outerRow := task.outerResult.GetRow(outerRowIdx)
+		for _, ptr := range innerRowPtrs {
+			matchedInnerRows = append(matchedInnerRows, task.innerResult.GetRow(ptr))
+		}
+		iter := chunk.NewIterator4Slice(matchedInnerRows)
+		for iter.Begin(); iter.Current() != iter.End(); {
+			matched, isNull, err := iw.joiner.tryToMatchInners(outerRow, iter, joinResult.chk)
+			if err != nil {
+				return err
+			}
+			hasMatched, hasNull = matched || hasMatched, isNull || hasNull
+			if joinResult.chk.IsFull() {
+				select {
+				case resultCh <- joinResult:
+				case <-ctx.Done():
+					return nil
+				}
+				joinResult, ok = iw.getNewJoinResult(ctx)
+				if !ok {
+					return errors.New("indexHashJoinInnerWorker.doJoinInOrder failed")
+				}
+			}
+		}
+		if !hasMatched {
+			iw.joiner.onMissMatch(hasNull, outerRow, joinResult.chk)
+		}
+	}
+	return nil
```

Comment:
Should we check and put the last `joinResult` into `resultCh` before return?

Reply:
It has been checked outside in `indexHashJoinInnerWorker.run`.

Reply:
Can the last `joinResult` be used by the next chunck? 
I think it is no because the the order of the next chunck may be not contiguous to the former one. But here is the next task  from the next chunck?